compliance/collection/nodeinventorizer/caching_inventorizer.go

```diff
@@ -0,0 +1,342 @@
+	if cachedInventory != nil && isCachedInventoryValid(creationTime) {
+		log.Debugf("Using cached node scan created at %v", creationTime)
+		return cachedInventory, nil
+	}
+
+	// check for existing backoff, wait for specified duration if needed, then persist the new backoff duration
+	initialBackoff := env.NodeScanInitialBackoff.DurationSetting()
+	currentBackoff := readBackoff(c.opts.BackoffFilePath)
+
+	if currentBackoff > initialBackoff {
+		log.Warnf("Found existing node scan backoff file - last scan may have failed. Waiting %v seconds before retrying", currentBackoff.Seconds())
+		c.opts.BackoffWaitCallback(currentBackoff)
+	}
+	writeBackoff(calcNextBackoff(currentBackoff), c.opts.BackoffFilePath)
+
+	// if no inventory exists, or it is too old, collect a fresh one and save it to the cache
+	newInventory, err := collectInventory(nodeName) // opts.Scanner.Scan(opts.NodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	err = writeInventory(newInventory, c.opts.InventoryCachePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "persisting inventory to cache")
+	}
+
+	// Remove backoff directly before returning message, so that a failing/killed container does not lead to
+	// frequent rescans of a Node, which are costly and might impact Node performance
+	removeBackoff(c.opts.BackoffFilePath)
+	return newInventory, nil
+}
+
+// readBackoff returns a backoff if found in given file, or the MaxBackoff on any error.
+//
+// It also checks the loaded backoff to not exceed the configured MaxBackoff.
+func readBackoff(path string) time.Duration {
+	backoff := env.NodeScanInitialBackoff.DurationSetting()
+	maxBackoff := env.NodeScanMaxBackoff.DurationSetting()
+
+	backoffFileContents, err := os.ReadFile(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No node scan backoff file found, continuing without pause")
+			return backoff
+		}
+		log.Warnf("Error reading node scan backoff file, continuing with MaxBackoff of %v. Error: %v", maxBackoff, err)
+		return maxBackoff
+	}
+
+	// We have an existing backoff counter
+	backoff, err = time.ParseDuration(string(backoffFileContents))
+	if err != nil {
+		log.Warnf("Error reading node scan backoff duration from file, continuing with MaxBackoff of %v. Error: %v", maxBackoff, err)
+		return maxBackoff
+	}
+
+	// Sanity-check the backoff loaded from disk
+	if backoff > maxBackoff {
+		log.Warnf("Existing node scan backoff exceeds MaxBackoff. Continuing with MaxBackoff of %v", maxBackoff)
+		return maxBackoff
+	}
+
+	return backoff
+}
+
+func writeBackoff(backoff time.Duration, path string) {
+	err := os.WriteFile(path, []byte(backoff.String()), 0644)
+	if err != nil {
+		log.Warnf("Error writing node scan backoff file: %v", err)
+	}
+}
+
+func calcNextBackoff(currentBackoff time.Duration) time.Duration {
+	maxBackoff := env.NodeScanMaxBackoff.DurationSetting()
+	nextBackoffInterval := currentBackoff + env.NodeScanBackoffIncrement.DurationSetting()
```

Comment:
We could also store these two in seconds as float64 and then the `calcNextBackoff` would be simply a `math.Min(maxBackoff, currentBackoff+backoffIncrement)`